## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
column_names = df.columns
print(column_names)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics
summary_stats = df.describe()
print(summary_stats)


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.



In [5]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()

top_categories = category_counts.head(3)

top_categories



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries,
# which subcategory had the most entries?
most_common_category = top_categories.index[0]

most_common_category_data = df[df['category'] == most_common_category]

subcategory_counts = most_common_category_data['subcategory'].value_counts()

most_common_subcategory = subcategory_counts.index[0]

most_common_subcategory


'bathroom supplies'

In [7]:
# Which five clients had the most entries in the data?
client_counts = df['client_id'].value_counts()

top_clients = client_counts.head(5)

top_clients


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.

top_client_ids = top_clients.index.tolist()

top_client_ids

[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the
# client with the most entries order order?

client_data = df[df['client_id'] == 33615]

total_units_ordered = client_data['qty'].sum()

total_units_ordered

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['subtotal'] = df['unit_price'] * df['qty']

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

df['total_weight'] = df['unit_weight'] * df['qty']

df['shipping_price'] = df['total_weight'].apply(lambda x: 7 if x > 50 else 10) * df['total_weight']

df[df['total_weight']>50]

first       last                                           job  \
0         Donald    Harding                                  Immunologist   
2        Shannon     Watson                                  Immunologist   
4      Christina   Schwartz                                  Chiropractor   
5        Jessica      Reyes  Production designer, theatre/television/film   
6          Brian      Moore                                       Barista   
...          ...        ...                                           ...   
54634     Lauren      Reese                                Radio producer   
54635    Derrick      Moore                                      Musician   
54636     Monica  Gutierrez                              Graphic designer   
54637      Wanda    Solomon                                  Toxicologist   
54638     Dwayne       Bell                       Chief Financial Officer   

                     phone                                email  client_id  \
0       793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
2        687.737.9424x8503                swatson8146@payne.net      57113   
4             265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5               5016915923             jessica.reyes@grimes.net      33615   
6      +1-893-400-2733x982          moore.official@martinez.com      87213   
...                    ...                                  ...        ...   
54634   977.877.5272x11382   late_reese_4081@montoya-chavez.org      29043   
54635         358.661.5483         derrick.moore.2602@pope.info      41908   
54636     294.805.9100x339            gutierrezm3195@morris.org      35176   
54637        (311)767-4924            solomonwanda5962@ross.org      24485   
54638    606-619-7667x9508                     belld@galvan.com      24007   

       order_id  order_date  order_week  order_year  ...     category  \
0       8953482  2023-04-28          17        2023  ...        decor   
2       1902144  2023-01-29           4        2023  ...     software   
4       1322274  2023-05-28          21        2023  ...  consumables   
5       1327730  2023-05-02          18        2023  ...        decor   
6       8080280  2023-04-19          16        2023  ...  consumables   
...         ...         ...         ...         ...  ...          ...   
54634   9021716  2023-04-26          17        2023  ...     software   
54635   6290153  2023-02-08           6        2023  ...  consumables   
54636   8692622  2023-03-05           9        2023  ...  consumables   
54637   7592730  2023-01-18           3        2023  ...    furniture   
54638   7489403  2023-01-04           1        2023  ...     software   

              subcategory unit_price  unit_cost  unit_weight   qty  \
0                wall art    1096.80     762.71         7.50   105   
2      project management      13.52       7.86         1.68    39   
4                    misc     195.10     108.17        46.43    20   
5                  plants     172.84     154.44         2.79    78   
6        kitchen supplies     245.56     231.95         1.73    56   
...                   ...        ...        ...          ...   ...   
54634           logistics      83.13      51.60         2.25    33   
54635    kitchen supplies     206.59     175.46        11.70    47   
54636   bathroom supplies      65.66      57.31         4.16   475   
54637              stands       1.48       1.22        18.04   112   
54638  project management       3.01       1.61         2.07  1031   

       line_number   subtotal  total_weight  shipping_price  
0                1  115164.00        787.50         5512.50  
2                6     527.28         65.52          458.64  
4                1    3902.00        928.60         6500.20  
5                0   13481.52        217.62         1523.34  
6                5   13751.36         96.88          678.16  
...            ...        ...           ...             ...

In [12]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

sales_tax_rate = 1.0925

df['total_price'] = round((df['subtotal'] + df['shipping_price']) * sales_tax_rate, 2)

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number   subtotal  total_weight  \
0             7.50   105            1  115164.00        787.50   
1             1.49    21            0     523.95         31.29   
2             1.68    39            6     527.28         65.52   
3             1.23    29            3    1056.18         35.67   
4            46.43    20            1    3902.00        928.60   
...            ...   ...          ...        ...           ...   
54634         2.25    33            8    2743.29         74.25   
54635        11.70    47            0    9709.73        549.90   
54636         4.16   475            6   31188.50       1976.00   
54637        18.04   112            0     165.76       2020.48   
54638         2.07  1031            5    3103.31       2134.17   

       shipping_price  total_price  
0             5512.50    131839.08  
1              312.90       914.26  
2              458.64      1077.12  
3              356.70      1543.57  
4             6500.20     11364.40  
...               ...          ...  
54634          519.75      3564.87  
54635         3849.30     14813.24  
54636        13832.00     49184.90  
54637        14143.36     15632.71  


In [13]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  total_weight  shipping_price  total_price  \
0                1  115164.00        787.50         5512.50    131839.08   
1                0     523.95         31.29          312.90       914.26   
2                6     527.28         65.52          458.64      1077.12   
3                3    1056.18         35.67          356.70      1543.57   
4                1    3902.00        928.60         6500.20     11364.40   
...            ...        ...           ...             ...          ...   
54634            8    2743.29         74.25          519.75      3564.87   
54635            0    9709.73        549.90         3849.30     14813.24   
54636            6   31188.50       1976.00        13832.00     49184.90   
54637            0     165.76       2020.48        14143.36     15632.71   
54638            5    3103.31       2134.17        14939.19     19711.43   

       line_cost  
0       85597.05  
1         629.79  
2         765.18  
3        1077.35  
4        8663.60  
...          ...  
54634    2222.55  
54635   12095.92  
54636   41054.25  
54637   14280.00  
54638   16599.10  

[54639 rows x 23 columns]

In [14]:
# Create a column for the profit of
# each line using line cost and line price

df['profit'] = df['total_price'] - df['line_cost']

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

        subtotal  total_weight  shipping_price  total_price  line_cost  \
0      115164.00        787.50         5512.50    131839.08   85597.05   
1         523.95         31.29          312.90       914.26     629.79   
2         527.28         65.52          458.64      1077.12     765.18   
3        1056.18         35.67          356.70      1543.57    1077.35   
4        3902.00        928.60         6500.20     11364.40    8663.60   
...          ...           ...             ...          ...        ...   
54634    2743.29         74.25          519.75      3564.87    2222.55   
54635    9709.73        549.90         3849.30     14813.24   12095.92   
54636   31188.50       1976.00        13832.00     49184.90   41054.25   
54637     165.76       2020.48        14143.36     15632.71   14280.00   
54638    3103.31       2134.17        14939.19     19711.43   16599.10   

         profit  
0      46242.03  
1        284.47  
2        311.94  
3        466.22  
4       2700.80  
...         ...  
54634   1342.32  
54635   2717.32  
54636   8130.65  
54637   1352.71  
54638   3112.33  

[54639 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above

order_2742071 = df[df['order_id'] == 2742071]
calculated_total_2742071 = order_2742071['total_price'].sum()
receipt_total_2742071 = 152811.89  # Total price from the receipt

print(calculated_total_2742071)
print(receipt_total_2742071)
print(calculated_total_2742071 == receipt_total_2742071)

#floating point rounding error

152811.88999999998
152811.89
False


In [16]:
order_2173913 = df[df['order_id'] == 2173913]
calculated_total_2173913 = order_2173913['total_price'].sum()
receipt_total_2173913 = 162388.71  # Total price from the receipt

print(calculated_total_2173913)
print(receipt_total_2173913)
print(calculated_total_2173913 == receipt_total_2173913)


162388.71
162388.71
True


In [17]:
order_6128929 = df[df['order_id'] == 6128929]
calculated_total_6128929 = order_6128929['total_price'].sum()
receipt_total_6128929 = 923441.25  # Total price from the receipt

print(calculated_total_6128929)
print(receipt_total_6128929)
print(calculated_total_6128929 == receipt_total_6128929)


923441.25
923441.25
True


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [56]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.


top_5_clients_spend_total = df.loc[df['client_id'].isin(top_clients.index), ['total_price', 'client_id']]

top_5_clients_spend_total.groupby('client_id')['total_price'].sum()



client_id
10033     689274.95
10091    1825072.92
10218    3683119.77
10351    1712403.23
10435    8722369.83
            ...    
54350    7057547.97
54400    3757304.89
54471    1706691.13
54518     489544.75
54596    2209398.96
Name: total_price, Length: 499, dtype: float64

In [63]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

top_clients_summary = top_clients.sort_values('profit', ascending=False)

top_clients['Total Revenue'] = top_clients['qty'] * top_clients['unit_price']

# Print the summary DataFrame
print(top_clients)


KeyError: 'profit'

In [64]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

top_clients = top_clients.rename(columns={
    'units_purchased': 'Total Units_Purchased',
    'shipping_price': 'Total Shipping_Price',
    'revenue': 'Total Revenue',
    'profit': 'Total Profit'
})

top_clients['Total Revenue'] = top_clients['Total Revenue'] / 1000000
top_clients['Total Profit'] = top_clients['Total Profit'] / 1000000

top_clients

first      last                                    job  \
21801    Julie  Anderson                    Exhibition designer   
15112   Dennis   Hawkins    Social research officer, government   
25972      Amy   Stewart              Psychologist, educational   
23929  Michele  Mcdonald  Emergency planning/management officer   
4063    Kendra   Garrett                 Pension scheme manager   
...        ...       ...                                    ...   
26881    Brian   Jackson                Human resources officer   
14104  Sabrina   Johnson                                 Artist   
1532    Kendra   Garrett                 Pension scheme manager   
35061     Seth  Thompson                   Intelligence analyst   
1830   Patrick  Espinoza                 Recruitment consultant   

                       phone                            email  client_id  \
21801             5933099462    anderson_better_4704@bell.biz      78965   
15112  001-214-888-5838x7410      hawkins.dennis.5195@lee.com      85931   
25972     979.793.4884x57205           stewartamy@francis.com      86721   
23929     435-931-4939x95137    mcdonaldm317@hess-jackson.com      38473   
4063            808.257.2029  garrettkendra8479@jefferson.com      24741   
...                      ...                              ...        ...   
26881    +1-697-510-2856x345  jackson_brian@edwards-gomez.com      71600   
14104   001-979-594-2130x182           johnsons6784@price.com      50002   
1532            808.257.2029  garrettkendra8479@jefferson.com      24741   
35061          (473)358-4887  thompsons3341@garcia-barber.com      47078   
1830              3533882109           pespinoza@alvarado.com      95108   

       order_id  order_date  order_week  order_year  ... unit_weight      qty  \
21801   3833568  2023-04-29          17        2023  ...        0.09  3958244   
15112   7263857  2023-03-05           9        2023  ...        2.21   637318   
25972   6057452  2023-03-21          12        2023  ...        9.22   977121   
23929   4594154  2023-03-02           9        2023  ...        2.25   219645   
4063    5167984  2023-05-08          19        2023  ...        1.86   122166   
...         ...         ...         ...         ...  ...         ...      ...   
26881   6822888  2023-02-03           5        2023  ...        2.88        0   
14104   5344708  2023-05-14          19        2023  ...        2.27        0   
1532    3335711  2023-04-15          15        2023  ...       12.91        0   
35061   7116918  2023-02-05           5        2023  ...       12.28        0   
1830    9744627  2023-05-30          22        2023  ...        6.31        0   

      line_number      subtotal  total_weight  Total Shipping Price  \
21801           1  5.844743e+08     356241.96            2493693.72   
15112           3  2.061214e+08    1408472.78            9859309.46   
25972           3  9.810295e+07    9009055.62           63063389.34   
23929           5  1.482472e+08     494201.25            3459408.75   
4063            7  5.635518e+07     227228.76            1590601.32   
...           ...           ...           ...                   ...   
26881           0  0.000000e+00          0.00                  0.00   
14104           3  0.000000e+00          0.00                  0.00   
1532            0  0.000000e+00          0.00                  0.00   
35061           0  0.000000e+00          0.00                  0.00   
1830            7  0.000000e+00          0.00                  0.00   

        total_price     line_cost  Total Profit  Total Revenue  
21801  6.412625e+08  3.987535e+08  2.425090e-10   5.844743e-10  
15112  2.359589e+08  1.270239e+08  1.089351e-10   2.061214e-10  
25972  1.760742e+08  1.344421e+08  4.163215e-11   9.810295e-11  
23929  1.657395e+08  1.282309e+08  3.750852e-11   1.482472e-10  
4063   6.330576e+07  3.188044e+07  3.142532e-11   5.635518e-11  
...             ...           ...           ...            ...  
26881  0.0

In [61]:
# Sort the updated data by "Total Profit" form highest to lowest

top_clients_summary = top_clients.sort_values('Total Profit', ascending=False)

# Print the summary DataFrame
print(top_clients_summary)

             first      last                                    job  \
21801        Julie  Anderson                    Exhibition designer   
15112       Dennis   Hawkins    Social research officer, government   
25972          Amy   Stewart              Psychologist, educational   
23929      Michele  Mcdonald  Emergency planning/management officer   
4063        Kendra   Garrett                 Pension scheme manager   
...            ...       ...                                    ...   
27866       Shelby  Williams                    Colour technologist   
21090       Daniel    Bowman                          Archaeologist   
21190        David    Herman                                 Artist   
13835  Christopher    Rhodes        Engineer, manufacturing systems   
1830       Patrick  Espinoza                 Recruitment consultant   

                       phone                                email  client_id  \
21801             5933099462        anderson_better_4704@bell.biz  